### PR TITLE
[GEP-28] Harmonize `gardenadm {init,bootstrap}` flows via new `flow.TaskGroup`

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity
/kind enhancement

**What this PR does / why we need it**:

`gardenadm init` and `gardenadm bootstrap` deploy substantially the same set of components (namespaces, cloud-provider secret, CRDs, `Cluster` resource, secrets management, `gardener-resource-manager`, extension controllers, network policies, infrastructure, OSC, MCM, worker, control plane, system components, …). Historically, each entry point assembled its own `flow.Graph` inline. This meant:

- The same component's deployment tasks — including dependency wiring, `SkipIf` conditions, retry policy, and task naming — were **written out twice**, once per entry point, with slight variations. Changes had to be applied to every occurrence, and drift crept in.
- **A component's own dependencies** (e.g., "GRM depends on secrets-management + CRDs + cluster-resource") lived at the caller rather than with the component's deployment logic, so the caller had to know these details.
- Once the graph was assembled, tasks that logically belonged together were **indistinguishable from their neighbours**: there was no "GRM" bundle to reference, only individual `TaskID`s.

To address this, this PR introduces `flow.TaskGroup` — a named, reusable bundle of tasks with a shared id and dependency set — and rewrites both flows to compose their graphs from shared groups defined once on `Botanist` / `GardenadmBotanist`. Each group encapsulates its component's deploy/wait tasks and its own upstream dependencies. Callers add groups with `g.AddGroup(...)`, optionally overlaying flow-specific `WithDependencies(...)` / `SkipIf(...)` at the call site. `Graph.AddGroup` flattens groups at build time — the compiled `Flow` has no notion of groups.

Flow-specific wiring stays at the caller (e.g., bootstrap's `syncPointBootstrapped` including network policies and extension controllers; init's task-level OSC handling downstream of Worker). This split keeps the shared surface honest: only genuinely-common structure lives in groups; per-flow specifics remain visible at the call site.

**Outlook**: the same task groups are the building blocks the gardenlet's shoot reconciliation will consume next, so the third caller falls into place without re-authoring the wiring.

**Which issue(s) this PR fixes**:
Part of #2906

**Release note**:
```feature developer
`flow.TaskGroup` is a new type in `pkg/utils/flow` for bundling tasks under a shared id and dependency set. Add it to a `flow.Graph` via `Graph.AddGroup`; the compiled `Flow` sees only individual tasks — groups are a build-time convenience. See the `godoc` on `TaskGroup` and `Graph.AddGroup` for the flattening rules.
```

### CODE REVIEW

**Comment by rfranzke** [comment]
/cc @ScheererJ 
FYI @timebertt

File: pkg/utils/flow/group.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	// When the group is added to a `Graph`, every task in the group inherits this set.
+	dependencies TaskIDs
+}
+
+// NewTaskGroup returns a new `TaskGroup` with the given id and the given initial tasks.
+func NewTaskGroup(id TaskID, tasks ...Task) TaskGroup {
+	return (TaskGroup{
+		id:           id,
+		tasks:        make(map[TaskID]Task, len(tasks)),
+		dependencies: make(TaskIDs, len(tasks)),
```
Comment (ScheererJ) [inline pkg/utils/flow/group.go]: Why do we make `depedencies` of length `tasks`? These seem to be very different entities. Please explain.

File: pkg/utils/flow/group.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+// Because resolution happens at `AddGroup` time, groups referenced by id via `WithDependencies`
+// must be added before the groups that depend on them; otherwise the id is treated as a plain
+// task id, which then panics for lack of a matching task.
+//
+// Like `Add`, this panics on duplicate task ids or unknown task-level dependencies.
+func (g *Graph) AddGroup(group TaskGroup) TaskIDs {
+	g.groups[group.id] = group
+
+	dependencies := make(TaskIDs)
+	for dependency := range group.dependencies {
```
Comment (ScheererJ) [inline pkg/utils/flow/group.go]: Not sure if we want to optimise this, but most other allocations include a length in contrast to this one. Assuming we rarely expect group dependencies we could make it `len(group.dependencies)`.
WDYT?

File: pkg/utils/flow/group.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+		}
+		if !progress {
+			// Every remaining task depends on another remaining task — either a cycle inside the group
+			// or a task-level dependency that names a task outside the group and outside the graph.
+			// Delegate to `g.Add` for one of them so it panics with its usual message.
+			for _, task := range pending {
+				g.Add(task)
+			}
+		}
+	}
```
Comment (ScheererJ) [inline pkg/utils/flow/group.go]: It might be just me, but I find these nested loops very complicated.
From what I understand, it seems the reason is that we do not want to add the tasks in an incorrect order and we store them in a map, which is not ordered. Would it not be easier to store the tasks in a slice in addition to the map so that we have the insertion order? This would allow us to simply iterate over the slice and insert tasks in the correct order assuming that they are specified in the correct order.
WDYT?

File: pkg/utils/flow/group_test.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+		extra := graph.Add(rec.task("extra"))
+		graph.AddGroup(
+			flow.NewTaskGroup("group", rec.task("a", extra)).WithDependencies(pre),
+		)
+
+		Expect(graph.Compile().Run(context.Background(), flow.Opts{})).To(Succeed())
+		executed := rec.executed()
+		Expect(executed).To(ContainElements("pre", "extra", "a"))
+		Expect(slices.Index(executed, "a")).To(BeNumerically(">", slices.Index(executed, "pre")))
+		Expect(slices.Index(executed, "a")).To(BeNumerically(">", slices.Index(executed, "extra")))
```
Comment (ScheererJ) [inline pkg/utils/flow/group_test.go]: While this is certainly possible you could also do analogously to lines 135/136, which seems easier to understand:

```suggestion
		Expect(executed[:2]).To(ConsistOf("pre", "extra"))
		Expect(executed[2]).To(Equal("a"))
```

File: pkg/gardenlet/operation/botanist/flow.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/gardener/gardener/pkg/utils/flow"
+	gardenerutils "github.com/gardener/gardener/pkg/utils/gardener"
+)
+
+// TaskGroupDeployNamespaces is a flow.TaskID for a logical flow.TaskGroup.
+const TaskGroupDeployNamespaces flow.TaskID = "TaskGroupDeployNamespaces"
+
+// DeployNamespacesTaskGroup returns the flow.TaskGroup for deploying the control plane namespace and the garden
+// namespace. The garden namespace is only deployed for self-hosted shoots.
+func (b *Botanist) DeployNamespacesTaskGroup() flow.TaskGroup {
```
Comment (ScheererJ) [inline pkg/gardenlet/operation/botanist/flow.go:34]: I assume that the remark concerning the garden namespace relates to the fact that it is only deployed as part of `gardenadm init` and not `gardenadm bootstrap`. However, using "only deployed for self-hosted shoots" to explain that might cause misunderstanding as `gardenadm bootstrap` also deploys a self-hosted shoot cluster after all. Maybe we should rephrase the last sentence of the comment.
WDYT?

File: pkg/gardenlet/operation/botanist/flow.go [hunk 6]
```diff
@@ -33,3 +33,16 @@ func (b *Botanist) DeployNamespacesTaskGroup() flow.TaskGroup {
...
+
+// DeployCloudProviderSecretTaskGroup returns the flow.TaskGroup for deploying the cloud provider secret. This task is
+// skipped when no shoot credentials are configured.
+func (b *Botanist) DeployCloudProviderSecretTaskGroup() flow.TaskGroup {
+	return flow.NewTaskGroup(TaskGroupDeployCloudProviderSecret, flow.Task{
+		Name:   "Deploying cloud provider account secret",
+		Fn:     b.DeployCloudProviderSecret,
+		SkipIf: b.Shoot.Credentials == nil,
+	}).WithDependencies(TaskGroupDeployNamespaces)
+}
```
Comment (ScheererJ) [inline pkg/gardenlet/operation/botanist/flow.go:61]: Just wondering about the usage of constants for the dependencies (`TaskGroupDeployNamespaces`). Previously, we always used the returned task id as dependency. We could continue to do so, but would need to pass them along as function parameter.
From my point of view, it makes sense to use the returned task id and pass it along to dependent tasks as this ensures that the registration order stays in sync, i.e. a task only depends on already registered task/task groups. With the use of constants you can easily break this.
WDYT?

File: pkg/gardenlet/operation/botanist/flow.go [hunk 7]
```diff
@@ -46,3 +55,55 @@ func (b *Botanist) DeployCloudProviderSecretTaskGroup() flow.TaskGroup {
 		SkipIf: b.Shoot.Credentials == nil,
 	}).WithDependencies(TaskGroupDeployNamespaces)
 }
+
+// TaskGroupReconcileCustomResourceDefinitions is a flow.TaskID for a logical flow.TaskGroup.
+const TaskGroupReconcileCustomResourceDefinitions flow.TaskID = "TaskGroupReconcileCustomResourceDefinitions"
+
+// ReconcileCustomResourceDefinitionsTaskGroup returns the flow.TaskGroup for reconciling the CRDs used by Gardener.
+// Each CRD is deployed in an individual task, and it is waited until the CRDs are marked as ready/available.
+func (b *Botanist) ReconcileCustomResourceDefinitionsTaskGroup() flow.TaskGroup {
```
Comment (ScheererJ) [inline pkg/gardenlet/operation/botanist/flow.go:68]: Is there a particular reason why `EnsureCustomResourceDefinitionsReady` no longer waits until the CRDs are healthy as was done in the past?
We also lost the corresponding tests as part of this. Did you not see any benefit in those?

File: pkg/gardenlet/operation/botanist/flow.go [hunk 8]
```diff
@@ -107,3 +108,16 @@ func (b *Botanist) ReconcileCustomResourceDefinitionsTaskGroup() flow.TaskGroup
...
+
+// ReconcileClusterResourceTaskGroup returns the flow.TaskGroup for reconciling the Cluster resource.
+func (b *Botanist) ReconcileClusterResourceTaskGroup() flow.TaskGroup {
+	return flow.NewTaskGroup(TaskGroupReconcileClusterResource, flow.Task{
+		Name: "Reconciling extensions.gardener.cloud/v1alpha1.Cluster resource",
+		Fn: func(ctx context.Context) error {
+			return gardenerextensions.SyncClusterResourceToSeed(ctx, b.SeedClientSet.Client(), b.Shoot.ControlPlaneNamespace, b.Shoot.GetInfo(), b.Shoot.CloudProfile, nil)
+		},
+	}).WithDependencies(TaskGroupReconcileCustomResourceDefinitions)
+}
```
Comment (ScheererJ) [inline pkg/gardenlet/operation/botanist/flow.go:126]: What is the reason behind using a `TaskGroup` for single tasks? Is it just uniformity or are there other reasons?

File: pkg/gardenadm/cmd/bootstrap/bootstrap.go [hunk 9]
```diff
@@ -109,29 +108,16 @@ func run(ctx context.Context, opts *Options) error {
...
-			Name: "Deploying seed system resources",
-			Fn: func(ctx context.Context) error {
-				return seedsystem.New(b.SeedClientSet.Client(), b.Shoot.ControlPlaneNamespace, seedsystem.Values{ManagePriorityClasses: true}).Deploy(ctx)
-			},
-			Dependencies: flow.NewTaskIDs(waitUntilGardenerResourceManagerReady),
-		})
+		reconcileGardenerResourceManager = g.AddGroup(
+			b.ReconcileGardenerResourceManagerTaskGroup(true, false).
+				WithDependencies(deployPriorityClassCritical),
+		)
```
Comment (ScheererJ) [inline pkg/gardenadm/cmd/bootstrap/bootstrap.go:115]: Why are the system resources now part of the `gardener-resource-manager` group? The naming of the group seems surprising.
That being said, building bigger groups may have benefits in reducing the overall flow size in the consumers (`bootstrap.go` and `init.go`).

File: pkg/gardenlet/operation/botanist/flow.go [hunk 10]
```diff
@@ -132,3 +133,67 @@ func (b *Botanist) InitializeSecretsManagementTaskGroup() flow.TaskGroup {
...
+			TaskGroupDeployNamespaces,
+			TaskGroupInitializeSecretsManagement,
+			TaskGroupReconcileCustomResourceDefinitions,
+		)
+		gardenadmBootstrap = b.Shoot.IsSelfHosted() && !b.Shoot.RunsControlPlane()
+
+		deployGardenerResourceManager = g.Add(flow.Task{
+			Name: "Deploying gardener-resource-manager",
+			Fn: func(ctx context.Context) error {
+				b.Shoot.Components.ControlPlane.RuntimeResourceManager.SetBootstrapControlPlaneNode(!podNetworkAvailable)
```
Comment (ScheererJ) [inline pkg/gardenlet/operation/botanist/flow.go]: Will this not crash in the `gardenadm bootstrap` case as `RuntimeResourceManager` only seems to be initialised in the `gardenadm init` case?
Moving this function call below the `if` block might do the trick.

File: pkg/gardenadm/botanist/flow.go [hunk 11]
```diff
@@ -37,3 +37,14 @@ func (b *GardenadmBotanist) ReconcileExtensionControllersTaskGroup(podNetworkAva
 
 	return g
 }
+
+// TaskGroupReconcileNetworkPolicies is a flow.TaskID for a logical flow.TaskGroup.
+const TaskGroupReconcileNetworkPolicies flow.TaskID = "TaskGroupReconcileNetworkPolicies"
+
+// ReconcileNetworkPoliciesTaskGroup returns the flow.TaskGroup for initializing the secret management.
+func (b *GardenadmBotanist) ReconcileNetworkPoliciesTaskGroup() flow.TaskGroup {
```
Comment (ScheererJ) [inline pkg/gardenadm/botanist/flow.go]:
```suggestion
// ReconcileNetworkPoliciesTaskGroup returns the flow.TaskGroup for reconciling the network policies.
func (b *GardenadmBotanist) ReconcileNetworkPoliciesTaskGroup() flow.TaskGroup {
```

File: pkg/gardenadm/cmd/bootstrap/bootstrap.go [hunk 12]
```diff
@@ -125,17 +125,10 @@ func run(ctx context.Context, opts *Options) error {
...
-		})
-		waitUntilOperatingSystemConfigReady = g.Add(flow.Task{
-			Name:         "Waiting until OperatingSystemConfig for control plane machines has been reconciled",
-			Fn:           b.Shoot.Components.Extensions.OperatingSystemConfig.Wait,
-			Dependencies: flow.NewTaskIDs(deployOperatingSystemConfig),
-		})
+		reconcileOperatingSystemConfig = g.AddGroup(
+			b.ReconcileInfrastructureTaskGroup().
+				WithDependencies(gardenadmbotanist.TaskGroupReconcileExtensionControllers, gardenadmbotanist.TaskGroupReconcileNetworkPolicies),
+		)
```
Comment (ScheererJ) [inline pkg/gardenadm/cmd/bootstrap/bootstrap.go]: Should this not be the OSC task group?

```suggestion
		reconcileOperatingSystemConfig = g.AddGroup(b.ReconcileOperatingSystemConfigTaskGroup())
```

File: pkg/gardenadm/cmd/bootstrap/bootstrap.go [hunk 13]
```diff
@@ -100,18 +100,10 @@ func run(ctx context.Context, opts *Options) error {
...
 
-		deployNamespaces                   = g.AddGroup(b.DeployNamespacesTaskGroup())
-		_                                  = g.AddGroup(b.DeployCloudProviderSecretTaskGroup())
-		reconcileCustomResourceDefinitions = g.Add(flow.Task{
-			Name: "Reconciling CustomResourceDefinitions",
-			Fn:   b.ReconcileCustomResourceDefinitions,
-		})
-		ensureCustomResourceDefinitionsReady = g.Add(flow.Task{
-			Name:         "Ensuring CustomResourceDefinitions are ready",
-			Fn:           flow.TaskFn(b.EnsureCustomResourceDefinitionsReady).RetryUntilTimeout(time.Second, time.Minute),
```
Comment (ScheererJ) [inline pkg/gardenadm/cmd/bootstrap/bootstrap.go:119]: We no longer seem to use the `RetryUntilTimeout` logic in the new code. Is that intentional?

File: pkg/gardenadm/botanist/flow.go [hunk 14]
```diff
@@ -6,30 +6,45 @@ package botanist
...
 	var (
-		g = flow.NewTaskGroup(TaskGroupReconcileExtensionControllers).WithDependencies(
-			botanist.TaskGroupReconcileGardenerResourceManager,
-		)
+		groupID, taskIDGardenerResourceManagerDependency = TaskGroupReconcileExtensionControllers, botanist.TaskGroupReconcileGardenerResourceManager
+		taskNameDeployment                               = "Deploying extension controllers"
+		taskNameWait                                     = "Waiting until extension controllers report readiness"
+	)
+	if podNetworkAvailable {
+		groupID, taskIDGardenerResourceManagerDependency = TaskGroupReconcileExtensionControllersInPodNetwork, botanist.TaskGroupReconcileGardenerResourceManagerInPodNetwork
```
Comment (ScheererJ) [inline pkg/gardenadm/botanist/flow.go]: Why not give `groupID` and `taskIDGardenerResourceManagerDependency` a separate line for each assignment? It looks surprising if only those two are on a single line.

**Review by ScheererJ — COMMENTED** [review]
Thanks for bringing `gardenadm init` and `gardenadm bootstrap` closer together and removing a lot of code duplication.

File: pkg/gardenlet/operation/botanist/flow.go (hunk 6 above)
Comment (rfranzke) [reply pkg/gardenlet/operation/botanist/flow.go:61]: I understand your point and also somewhat agree to it: If a constant dependency (like this) is not registered in the flow graph you're constructing, it will panic. The good thing is: It will already panic when the graph is compiled (`Compile()` method), so you'll quickly notice it during development.

The motivation of the constant dependencies is that you don't have to wire them up again and again in each usage of the `TaskGroup`s in your flow graphs. This is part of the motivation of the entire change: We want to make common things as reusable as possible.

File: pkg/gardenlet/operation/botanist/flow.go (hunk 7 above)
Comment (rfranzke) [reply pkg/gardenlet/operation/botanist/flow.go:68]: There is no `EnsureCustomResourceDefinitionsReady` function anymore. This `ReconcileCustomResourceDefinitionsTaskGroup` directly uses `component.OpWait()` with the CRD deployer component to wait for their readiness. Hence, functionality should be preserved (and even more precise now - we only wait for the CRDs we were applying - previously, `EnsureCustomResourceDefinitionsReady` waited for all CRDs).

File: pkg/gardenadm/cmd/bootstrap/bootstrap.go (hunk 9 above)
Comment (rfranzke) [reply pkg/gardenadm/cmd/bootstrap/bootstrap.go:115]: Agree - this is already split up again in a later commit in this PR.

File: pkg/gardenadm/cmd/bootstrap/bootstrap.go (hunk 13 above)
Comment (rfranzke) [reply pkg/gardenadm/cmd/bootstrap/bootstrap.go:119]: Yes - we use `component.OpWait` which actively blocks while the CRD is not ready. Previously, `EnsureCustomResourceDefinitionsReady` wasn't blocking but just did a single check.

File: pkg/gardenadm/botanist/flow.go (hunk 14 above)
Comment (rfranzke) [reply pkg/gardenadm/botanist/flow.go]: No strong reason - personal preference in the moment I was typing this. I can also separate them 😅

File: pkg/gardenlet/operation/botanist/flow.go (hunk 8 above)
Comment (rfranzke) [reply pkg/gardenlet/operation/botanist/flow.go:126]: Indeed, uniformity, plus the group id serves as a stable dependency handle across the package - `WithDependencies(TaskGroupXxx)` works whether a group has one task or many, without callers having to capture `TaskID`s. Also keeps expansion (e.g., adding a `Wait` task later) transparent for callers.

File: pkg/gardenlet/operation/botanist/flow.go (hunk 6 above)
Comment (ScheererJ) [reply pkg/gardenlet/operation/botanist/flow.go:61]: On the common part, I completely agree that changes will likely cause `panic`s directly and hence the first manual test of a flawed code change will make it clearly visible that something is broken. Unfortunately, we also have code paths which are not executed all the time, e.g. if `skipReadiness` or some other edge case is used.
That being said, we can start with this approach and accept the risk for now.

File: pkg/gardenlet/operation/botanist/flow.go (hunk 7 above)
Comment (ScheererJ) [reply pkg/gardenlet/operation/botanist/flow.go:68]: Ok, let me rephrase.
Why do we no longer perform the health check of the CRDs (`health.CheckCustomResourceDefinition(&crd)`) as we did before? Indeed we apply the CRDs and wait for that to be completed, but previously we also waiting for the CRDs to be healthy, which requires another controller to act on the CRDs.
I know that this has caused (test) issues in the past. Was this an intentional removal/omission in the new code?

File: pkg/gardenlet/operation/botanist/flow.go (hunk 8 above)
Comment (ScheererJ) [reply pkg/gardenlet/operation/botanist/flow.go:126]: Well, a single task can also have a very stable task identifier, but I can understand the uniformity aspect of it.

File: pkg/gardenlet/operation/botanist/flow.go (hunk 7 above)
Comment (rfranzke) [reply pkg/gardenlet/operation/botanist/flow.go:68]: I don't think we changed anything - the `Wait` function of the `crddeployer` package runs the same `health. CheckCustomResourceDefinition` function, i.e., it only returns when the CRD is healthy. Maybe I'm still missing something?
